### DISCUSSION TITLE
What are the differences between React18 Streaming SSR  and islands architecture?

### DISCUSSION THREAD

**Discussion by zwl1619** [vercel/next.js#47348] [Help]
### Summary

What are the differences between React18 Streaming SSR  and islands architecture?

**Comment by icyJoseph** [comment]
I think it is best to establish some literature first:

- https://github.com/reactwg/react-18/discussions/37
- https://www.patterns.dev/posts/islands-architecture

From the **React Working Group**:

> This is not a novel idea: for example, [Marko](https://tech.ebayinc.com/engineering/async-fragments-rediscovering-progressive-html-rendering-with-marko/) is one of JavaScript web frameworks that implements a version of this pattern. The challenge was in how to adapt a pattern like this to the React programming model. It took a while to solve. We introduced the <Suspense> component for this purpose in 2018. When we introduced it, we only supported it for lazy-loading code on the client. But the goal was to integrate it with the server rendering and solve these problems.

So if we think of **Islands Architecture** beyond the Astro framework, and rather, understand it as a meta-technique by which we can send partition the application into static, data-dependent and interactive sections, we can see that the **React 18** technique is trying to be but an implementation of such meta-technique.

Strictly speaking, **React 18 Streaming SSR** falls somewhere, between **Progressive Hydration** and the **Islands Architecture**. However, **React Server Components** do bridge that gap, by just straight up rendering components server side, which generates static HTML that in turn can wrap (or be wrapped by) the so called **Client Components**, which are responsible for interactivity.

It is usually this case with these innovative technologies, they keep blurring boundaries and escaping definitions, labels/tags we put on them are prone to become stale quickly.

Of course, how we build mental models around this concepts is also very much plastic, I feel like I might come back to this later, and not agree with myself, we'll see.

**Reply by stephanrauh** [reply]
@icyJoseph Can you give us an update? Has next.js moved closer to the Islands Architecture since you answered the question exactly one year ago?

Currently, I'm following the next.js tutorial, and when I saw the section about partial prerendering, I thought "that's the Islands Architecture". However, when I ran the demo, it doesn't send simple HTML files to the client. I saw a lot of JavaScript, and an HTML file that finishes loading when every server-side component has finished loading. The screenshot exaggerates it a bit because I've deliberately added a delay.

<img width="1131" alt="image" src="https://github.com/vercel/next.js/assets/3045767/a4fdd1c2-ce38-47a7-b590-c182b58b246c">

I have to admit that the page renders virtually immediately and fills in some gaps later. From a user's point of view, that's perfect.

However, the idea of the islands architecture is to send HTML first, so search engines can read the page without having to interpret JavaScript. I wonder what a search engine does when it runs into an HTML page that takes 3 seconds. I'm afraid it assumes the page loads slow, resulting in a bad page rank.

Did I miss something? Can I configure next.js so it sends small HTML files immediately?

Just in case you want to look at my sourcecode: https://github.com/stephanrauh/ExploringNextJS/tree/main/nextjs-dashboard
